### ISSUE TITLE
opt: Extract restrictions from OR of AND clauses in join conditions

### ISSUE THREAD

**Issue by andy-kimball** [cockroachdb/cockroach#38721]
Sometimes a join condition contains an OR clause that references both sides of the join and therefore cannot be pushed down to either side in order to reduce input cardinalities. However, it is sometimes possible to extract restrictions that can be pushed down, as in this example from TPC-H Q7:
```
AND (
    (n1.n_name = 'FRANCE' AND n2.n_name = 'GERMANY')
    or (n1.n_name = 'GERMANY' AND n2.n_name = 'FRANCE')
)
```
Here, the following restrictions can be extracted (and duplicated):
```
AND (
    (n1.n_name = 'FRANCE' AND n2.n_name = 'GERMANY')
    or (n1.n_name = 'GERMANY' AND n2.n_name = 'FRANCE')
)
AND n1.n_name IN ('FRANCE', 'GERMANY')
AND n2.n_name IN ('FRANCE', 'GERMANY')
```
Since these new restrictions are now isolated to one input or the other, they can be pushed down and reduce the number of rows that need to be fed into the join.

Postgres does this optimization. See here:
https://github.com/postgres/postgres/blob/8255c7a5eeba8f1a38b7a431c04909bde4f5e67d/src/backend/optimizer/util/orclauses.c

When I apply this optimization manually to TPC-H Q7, it reduces the performance of that query from ~8s to ~4s on my laptop.

**Comment by andy-kimball** [comment]
CC @rytaft, who has added some AND/OR optimizations similar to this in the past.

**Comment by RaduBerinde** [comment]
FWIW, the `splitFilter` primitive used by the HP can do this.
https://github.com/cockroachdb/cockroach/blob/2eebbddbb133eea7102a47fbe7f5d13ec6f8f670/pkg/sql/expr_filter.go#L387
https://github.com/cockroachdb/cockroach/blob/2eebbddbb133eea7102a47fbe7f5d13ec6f8f670/pkg/sql/expr_filter_test.go#L90

We might need to implement something similar and use it in the join rules.

**Comment by rytaft** [comment]
Seems like we could handle this case pretty easily with a rule similar to `ExtractRedundantConjunct`: https://github.com/cockroachdb/cockroach/blob/78bcb5a88af823734c90124c720ed6e3180d4a96/pkg/sql/opt/norm/rules/bool.opt#L157. Instead of checking that the conjunct matches exactly in all clauses, we would just need to check that there is an equality expression with matching outer columns.

**Comment by rytaft** [comment]
Just to add an update here: I've investigated adding a rule similar to `ExtractRedundantConjunct`. As I suspected it is relatively easy to do, but there is a problem: since the original filter remains intact after the rule is applied, it's not currently possible to tell that the rule has already been applied. As a result, the rule may be applied over and over again in an infinite loop.

In a separate context we have discussed "bubbling up" constraints throughout the query tree. This will be necessary to optimize queries like `SELECT * FROM (SELECT * FROM a WHERE x=5) AS a JOIN b ON a.x=b.x;`. In this example, if we "bubble up" the filter `a.x=5` as a constraint `/x: [/5 - /5]`, then using the equality `a.x=b.x` we can also generate and push down the filter `b.x=5`.

The infrastructure needed to support bubbling up constraints and pushing down filters will fix this issue and will solve the problem with the infinite loop. Putting this issue on hold for the time being until we build the infrastructure to bubble up constraints.

**Comment by DrewKimball** [comment]
@rytaft if you still have that branch, a temporary solution for this release could be to use the UnfilteredCols ruleprop in lieu of bubbling up contraints; it might be a nice way to tap some extra TPC-H performance.

**Comment by rytaft** [comment]
That's an interesting idea... so to avoid the infinite loop, the rule should check the UnfilteredCols prop, and only fire if that prop is unset?

**Comment by DrewKimball** [comment]
Yes. It wouldn't handle as many cases as bubbling up constraints, but I think it should handle Q7.

**Comment by andy-kimball** [comment]
@rytaft, could you point @DrewKimball to the branch? He could take it over and implement this.

**Comment by rytaft** [comment]
I believe this is the branch that I was working on: https://github.com/rytaft/cockroach/tree/extract-or-clause

Let me know if I can help explain anything, since I obviously never got it fully documented or into commit-ready shape

**Comment by DrewKimball** [comment]
I'm not seeing any difference in runtime, so it probably isn't worth trying to fit this into this release.

**Comment by rytaft** [comment]
Too bad -- if you get a chance, can you post your branch here so maybe I can try to dig into this later (unless you think you'll continue to work on it)?

**Comment by DrewKimball** [comment]
Here it is: https://github.com/DrewKimball/cockroach/tree/extract-or-clause
